# Library imports and data loading

In [1]:
# Python verstion - 3.10.8
#%pip install pandas

In [2]:
import pandas as pd
import re

In [3]:
df = pd.read_csv('data/Consumer_Airfare_Report__Table_1a_-_All_U.S._Airport_Pair_Markets_20240712.csv')

C:\Users\FTW\AppData\Local\Temp\ipykernel_9940\1818994699.py:1: DtypeWarning: Columns (20,21) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('data/Consumer_Airfare_Report__Table_1a_-_All_U.S._Airport_Pair_Markets_20240712.csv')


# Data exploration

In [4]:
print(df.shape)
df.head()

(245955, 23)


tbl  Year  quarter  citymarketid_1  citymarketid_2  \
0  Table1a  2021        3           30135           33195   
1  Table1a  2021        3           30135           33195   
2  Table1a  2021        3           30140           30194   
3  Table1a  2021        3           30140           30194   
4  Table1a  2021        3           30140           30466   

                            city1                          city2  airportid_1  \
0  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)        10135   
1  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)        10135   
2                 Albuquerque, NM          Dallas/Fort Worth, TX        10140   
3                 Albuquerque, NM          Dallas/Fort Worth, TX        10140   
4                 Albuquerque, NM                    Phoenix, AZ        10140   

   airportid_2 airport_1  ...    fare  carrier_lg  large_ms  fare_lg  \
0        14112       ABE  ...   81.43          G4    1.0000    81.43   
1        15304       ABE  ...  208.93          DL    0.4659   219.98   
2        11259       ABQ  ...  184.56          WN    0.9968   184.44   
3        11298       ABQ  ...  182.64          AA    0.9774   183.09   
4        14107       ABQ  ...  177.11          WN    0.6061   184.49   

  carrier_low   lf_ms  fare_low Geocoded_City1  Geocoded_City2  \
0          G4  1.0000     81.43            NaN             NaN   
1          UA  0.1193    154.11            NaN             NaN   
2          WN  0.9968    184.44            NaN             NaN   
3          AA  0.9774    183.09            NaN             NaN   
4          AA  0.3939    165.77            NaN             NaN   

                 tbl1apk  
0  202131013514112ABEPIE  
1  202131013515304ABETPA  
2  202131014011259ABQDAL  
3  202131014011298ABQDFW  
4  202131014014107ABQPHX  

[5 rows x 23 columns]

In [5]:
df[['Year','quarter','nsmiles','passengers','fare']].describe()

Year        quarter        nsmiles     passengers  \
count  245955.000000  245955.000000  245955.000000  245955.000000   
mean     2008.524124       2.479153    1189.812319     299.476795   
std         8.703364       1.122149     703.143472     511.389486   
min      1993.000000       1.000000     109.000000       0.000000   
25%      2001.000000       1.000000     626.000000      21.000000   
50%      2008.000000       2.000000    1023.000000     113.000000   
75%      2016.000000       3.000000    1736.000000     339.000000   
max      2024.000000       4.000000    2724.000000    8301.000000   

                fare  
count  245955.000000  
mean      218.979587  
std        82.372486  
min        50.000000  
25%       164.620000  
50%       209.320000  
75%       262.890000  
max      3377.000000

In [6]:
df.describe(include=['O'])

tbl                                city1  \
count     245955                               245955   
unique         2                                  141   
top     Table 1a  Los Angeles, CA (Metropolitan Area)   
freq      200897                                26055   

                                        city2 airport_1 airport_2 carrier_lg  \
count                                  245955    245955    245955     244415   
unique                                    128       187       173         67   
top     New York City, NY (Metropolitan Area)       ORD       DCA         WN   
freq                                    36913      8819     10268      58999   

       carrier_low                                     Geocoded_City1  \
count       244343                                             206749   
unique          71                                                238   
top             WN  Los Angeles, CA (Metropolitan Area)\n(34.05223...   
freq         56027                                              21028   

                                           Geocoded_City2  \
count                                              206749   
unique                                                207   
top     New York City, NY (Metropolitan Area)\n(40.123...   
freq                                                30688   

                      tbl1apk  
count                  245955  
unique                 245955  
top     202411541215304TYSTPA  
freq                        1

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 245955 entries, 0 to 245954
Data columns (total 23 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   tbl             245955 non-null  object 
 1   Year            245955 non-null  int64  
 2   quarter         245955 non-null  int64  
 3   citymarketid_1  245955 non-null  int64  
 4   citymarketid_2  245955 non-null  int64  
 5   city1           245955 non-null  object 
 6   city2           245955 non-null  object 
 7   airportid_1     245955 non-null  int64  
 8   airportid_2     245955 non-null  int64  
 9   airport_1       245955 non-null  object 
 10  airport_2       245955 non-null  object 
 11  nsmiles         245955 non-null  int64  
 12  passengers      245955 non-null  int64  
 13  fare            245955 non-null  float64
 14  carrier_lg      244415 non-null  object 
 15  large_ms        244415 non-null  float64
 16  fare_lg         244415 non-null  float64
 17  carrier_lo

In [8]:
df.isnull().sum()

tbl                   0
Year                  0
quarter               0
citymarketid_1        0
citymarketid_2        0
city1                 0
city2                 0
airportid_1           0
airportid_2           0
airport_1             0
airport_2             0
nsmiles               0
passengers            0
fare                  0
carrier_lg         1540
large_ms           1540
fare_lg            1540
carrier_low        1612
lf_ms              1612
fare_low           1612
Geocoded_City1    39206
Geocoded_City2    39206
tbl1apk               0
dtype: int64

## About the dataset

There are `245955 rows` and `23 columns` in the dataset. The data contains information about airline routes in US for the time period of 1993-2024. Each year contains 4 quarters. The attributes of the dataset include the city and airport (id and name) for depature and arrival. The dataset also includes columns `passengers` and `fare` which, according to the description of the dataset on Kaggle, are average number of passengers and average fare amount for each airline route presumabily for each quater (there was no comment about it). It also has brief market related data for each attribute such as the largest and "lowest" carrier (company) with their respective market share. Lastly, it has latitude and longitude for depature and arrival airports. `tbl1apk` column acts as a unique id for every single route of each year and quarter. There are a lot of missing values in a number of columns. We will investigate and take a closer look later.

## Next step in refining the data

There are a few things to be done in the dataset to "clean" it. 

- We should first rename the `passenger` and `fare` columns to reflect that they are average. 
- We will drop the `tbl` column as it provides no information. It is most likely just a column generated by the web scraper or some sort.
- Investigate each column with missing value and refine them. While this is a very generic description of what is needed to be done, exploration of each column would rather be documented when we get to each column as the missing values might not be uniform.
- For Geocode columns, there is some inconsistency in how some rows are documented. Basically, these columns contain latitude and longitude of the airports but the text around the lat and long values can be inconsistent. We may have to use regex to extract them out.
- Upon checking the `nsmiles` column, which contains miles flown by the plane for that particular route, we found that there are some inconsistent values. For example, flight flown from airport id 10135 to 14112 is noted as 970 miles for a lot of the rows but some rows are noted as 971 miles. This is likely to be the inconsistency generated by rounding up/down during the data generation process. We will have to check if there are any other case in the column and fix them. This is the only possible column with such issue since the distance is going to be fixed despite the year or quarter change.
- Similar to the last point, we ought to check if there are any syntactical errors in other columns.

Drop the `tbl` column

In [9]:
df = df.drop(columns =['tbl'])

Renaming `passengers` and `fare` columns

In [10]:
df.rename(columns = {'passengers':'avg_daily_passengers', 'fare': 'avg_fare'}, inplace = True)

In [11]:
df.head()

Year  quarter  citymarketid_1  citymarketid_2  \
0  2021        3           30135           33195   
1  2021        3           30135           33195   
2  2021        3           30140           30194   
3  2021        3           30140           30194   
4  2021        3           30140           30466   

                            city1                          city2  airportid_1  \
0  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)        10135   
1  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)        10135   
2                 Albuquerque, NM          Dallas/Fort Worth, TX        10140   
3                 Albuquerque, NM          Dallas/Fort Worth, TX        10140   
4                 Albuquerque, NM                    Phoenix, AZ        10140   

   airportid_2 airport_1 airport_2  ...  avg_fare  carrier_lg  large_ms  \
0        14112       ABE       PIE  ...     81.43          G4    1.0000   
1        15304       ABE       TPA  ...    208.93          DL    0.4659   
2        11259       ABQ       DAL  ...    184.56          WN    0.9968   
3        11298       ABQ       DFW  ...    182.64          AA    0.9774   
4        14107       ABQ       PHX  ...    177.11          WN    0.6061   

  fare_lg  carrier_low   lf_ms fare_low  Geocoded_City1  Geocoded_City2  \
0   81.43           G4  1.0000    81.43             NaN             NaN   
1  219.98           UA  0.1193   154.11             NaN             NaN   
2  184.44           WN  0.9968   184.44             NaN             NaN   
3  183.09           AA  0.9774   183.09             NaN             NaN   
4  184.49           AA  0.3939   165.77             NaN             NaN   

                 tbl1apk  
0  202131013514112ABEPIE  
1  202131013515304ABETPA  
2  202131014011259ABQDAL  
3  202131014011298ABQDFW  
4  202131014014107ABQPHX  

[5 rows x 22 columns]

Dealing with missing values

In [12]:
## TODO: Find out what missing values in market share (largest and lowest) are about

In [13]:
# subsetting a temp dataframe to check the missing values in geocode columns
temp = df[(df['airportid_1'] == 10135) & (df['airportid_2'] == 14112)]

In [14]:
temp.head()

Year  quarter  citymarketid_1  citymarketid_2  \
0       2021        3           30135           33195   
6030    2022        2           30135           33195   
39380   2008        1           30135           33195   
115953  2007        1           30135           33195   
167868  2007        3           30135           33195   

                                 city1                          city2  \
0       Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)   
6030    Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)   
39380   Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)   
115953  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)   
167868  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)   

        airportid_1  airportid_2 airport_1 airport_2  ...  avg_fare  \
0             10135        14112       ABE       PIE  ...     81.43   
6030          10135        14112       ABE       PIE  ...     98.33   
39380         10135        14112       ABE       PIE  ...    108.71   
115953        10135        14112       ABE       PIE  ...    104.83   
167868        10135        14112       ABE       PIE  ...     91.23   

        carrier_lg  large_ms fare_lg  carrier_low  lf_ms fare_low  \
0               G4       1.0   81.43           G4    1.0    81.43   
6030            G4       1.0   98.33           G4    1.0    98.33   
39380           G4       1.0  108.71           G4    1.0   108.71   
115953          G4       1.0  104.83           G4    1.0   104.83   
167868          G4       1.0   91.23           G4    1.0    91.23   

                                           Geocoded_City1  \
0                                                     NaN   
6030                                                  NaN   
39380   Allentown/Bethlehem/Easton, PA\n(40.602753, -7...   
115953  Allentown/Bethlehem/Easton, PA\n(40.602753, -7...   
167868  Allentown/Bethlehem/Easton, PA\n(40.602753, -7...   

                                           Geocoded_City2  \
0                                                     NaN   
6030                                                  NaN   
39380   Tampa, FL (Metropolitan Area)\n(37.8606, -78.8...   
115953  Tampa, FL (Metropolitan Area)\n(37.8606, -78.8...   
167868  Tampa, FL (Metropolitan Area)\n(37.8606, -78.8...   

                      tbl1apk  
0       202131013514112ABEPIE  
6030    202221013514112ABEPIE  
39380   200811013514112ABEPIE  
115953  200711013514112ABEPIE  
167868  200731013514112ABEPIE  

[5 rows x 22 columns]

In [15]:
#airportid_1 has 188 entries whereas airport_1 only has 187 values. One of the airports have 2 ids for some reason.
double_airport = df.groupby("airport_1")['airportid_1'].nunique()
ununique_airport = double_airport[double_airport > 1]
print(ununique_airport)
#Austin has 2 airport ids because one is the now defunct Robert Mueller Municipal Airport (closed 1999) and the other is Austin–Bergstrom International Airport.
#Robert Mueller Airport is 1556 miles from Long Island, as opposed to the 1559 miles from Austin-Bergstrom.

airport_1
AUS    2
Name: airportid_1, dtype: int64


In [16]:
## TODO: Check if there are any routes with ABSOLUTELY missing geocode values

# Finding out which routes have absolutely no geocode values
# Group by departure and arrival airport IDs and filter groups where all geocode values are NaN
consistently_nan_routes = df.groupby(['airportid_1', 'airportid_2']).filter(
    lambda x: x['Geocoded_City1'].isna().all() and x['Geocoded_City2'].isna().all()
)

In [17]:
print(consistently_nan_routes.shape[0])
consistently_nan_routes

334


Year  quarter  citymarketid_1  citymarketid_2  \
68      2021        3           30194           31135   
69      2021        3           30194           31135   
310     2021        3           30466           30198   
484     2021        3           30713           31703   
485     2021        3           30713           31703   
...      ...      ...             ...             ...   
245525  2024        1           32467           34685   
245526  2024        1           32467           34685   
245711  2024        1           32575           34314   
245712  2024        1           32575           34314   
245787  2024        1           33195           35356   

                                      city1  \
68                    Dallas/Fort Worth, TX   
69                    Dallas/Fort Worth, TX   
310                             Phoenix, AZ   
484                               Boise, ID   
485                               Boise, ID   
...                                     ...   
245525        Miami, FL (Metropolitan Area)   
245526        Miami, FL (Metropolitan Area)   
245711  Los Angeles, CA (Metropolitan Area)   
245712  Los Angeles, CA (Metropolitan Area)   
245787        Tampa, FL (Metropolitan Area)   

                                        city2  airportid_1  airportid_2  \
68                           Myrtle Beach, SC        11259        13577   
69                           Myrtle Beach, SC        11298        13577   
310                            Pittsburgh, PA        10466        14122   
484     New York City, NY (Metropolitan Area)        10713        11618   
485     New York City, NY (Metropolitan Area)        10713        12197   
...                                       ...          ...          ...   
245525                           Savannah, GA        11697        14685   
245526                           Savannah, GA        13303        14685   
245711                              Provo, UT        12892        14314   
245712                              Provo, UT        14908        14314   
245787                            Trenton, NJ        15304        15356   

       airport_1 airport_2  ...  avg_fare  carrier_lg  large_ms fare_lg  \
68           DAL       MYR  ...    167.45          WN    0.8949  163.64   
69           DFW       MYR  ...    166.15          AA    0.8389  172.96   
310          AZA       PIT  ...     60.14          G4    1.0000   60.14   
484          BOI       EWR  ...    329.44          UA    0.4573  388.71   
485          BOI       HPN  ...    319.75          UA    0.6000  366.63   
...          ...       ...  ...       ...         ...       ...     ...   
245525       FLL       SAV  ...    182.14          DL    0.4416  225.09   
245526       MIA       SAV  ...    259.37          AA    0.9200  254.50   
245711       LAX       PVU  ...     65.67          G4    1.0000   65.67   
245712       SNA       PVU  ...     87.64          MX    0.7316   90.03   
245787       TPA       TTN  ...    100.14          F9    1.0000  100.14   

        carrier_low   lf_ms fare_low  Geocoded_City1  Geocoded_City2  \
68               WN  0.8949   163.64             NaN             NaN   
69               NK  0.0881    87.79             NaN             NaN   
310              G4  1.0000    60.14             NaN             NaN   
484              DL  0.3626   284.14             NaN             NaN   
485              DL  0.3333   218.60             NaN             NaN   
...             ...     ...      ...             ...             ...   
245525           AA  0.4072   157.31             NaN             NaN   
245526           AA  0.9200   254.50             NaN             NaN   
245711           G4  1.0000    65.67             NaN             NaN   
245712           G4  0.2684    81.14             NaN             NaN   
245787           F9  1.0000   100.14             NaN             NaN   

                      tbl1apk  
68      202131125913577DALMYR  
69      202131129813577DFWM

In [18]:
# Find how many unique airline routes have missing geocode values
unique_geo_nan_routes = consistently_nan_routes[['airportid_1', 'airportid_2']].drop_duplicates()
print(unique_geo_nan_routes.shape[0])

126


In [19]:
#create routes column for grouping later
df["route"] = df["airport_1"] + "-" + df["airport_2"]
df.head()

Year  quarter  citymarketid_1  citymarketid_2  \
0  2021        3           30135           33195   
1  2021        3           30135           33195   
2  2021        3           30140           30194   
3  2021        3           30140           30194   
4  2021        3           30140           30466   

                            city1                          city2  airportid_1  \
0  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)        10135   
1  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)        10135   
2                 Albuquerque, NM          Dallas/Fort Worth, TX        10140   
3                 Albuquerque, NM          Dallas/Fort Worth, TX        10140   
4                 Albuquerque, NM                    Phoenix, AZ        10140   

   airportid_2 airport_1 airport_2  ...  carrier_lg  large_ms  fare_lg  \
0        14112       ABE       PIE  ...          G4    1.0000    81.43   
1        15304       ABE       TPA  ...          DL    0.4659   219.98   
2        11259       ABQ       DAL  ...          WN    0.9968   184.44   
3        11298       ABQ       DFW  ...          AA    0.9774   183.09   
4        14107       ABQ       PHX  ...          WN    0.6061   184.49   

  carrier_low   lf_ms  fare_low Geocoded_City1  Geocoded_City2  \
0          G4  1.0000     81.43            NaN             NaN   
1          UA  0.1193    154.11            NaN             NaN   
2          WN  0.9968    184.44            NaN             NaN   
3          AA  0.9774    183.09            NaN             NaN   
4          AA  0.3939    165.77            NaN             NaN   

                 tbl1apk    route  
0  202131013514112ABEPIE  ABE-PIE  
1  202131013515304ABETPA  ABE-TPA  
2  202131014011259ABQDAL  ABQ-DAL  
3  202131014011298ABQDFW  ABQ-DFW  
4  202131014014107ABQPHX  ABQ-PHX  

[5 rows x 23 columns]

In [20]:
unique_routes_depature = df[['route', 'Geocoded_City1']].dropna().drop_duplicates(subset='route')
unique_routes_arrival = df[['route', 'Geocoded_City2']].dropna().drop_duplicates(subset='route')

df = df.merge(unique_routes_depature, on='route', how='left', suffixes=('', '_fixed'))
df = df.merge(unique_routes_arrival, on='route', how='left',  suffixes=('', '_fixed'))
df.drop(columns=['Geocoded_City1', 'Geocoded_City2'], inplace=True)

In [21]:
df.head()

Year  quarter  citymarketid_1  citymarketid_2  \
0  2021        3           30135           33195   
1  2021        3           30135           33195   
2  2021        3           30140           30194   
3  2021        3           30140           30194   
4  2021        3           30140           30466   

                            city1                          city2  airportid_1  \
0  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)        10135   
1  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)        10135   
2                 Albuquerque, NM          Dallas/Fort Worth, TX        10140   
3                 Albuquerque, NM          Dallas/Fort Worth, TX        10140   
4                 Albuquerque, NM                    Phoenix, AZ        10140   

   airportid_2 airport_1 airport_2  ...  carrier_lg  large_ms  fare_lg  \
0        14112       ABE       PIE  ...          G4    1.0000    81.43   
1        15304       ABE       TPA  ...          DL    0.4659   219.98   
2        11259       ABQ       DAL  ...          WN    0.9968   184.44   
3        11298       ABQ       DFW  ...          AA    0.9774   183.09   
4        14107       ABQ       PHX  ...          WN    0.6061   184.49   

  carrier_low   lf_ms  fare_low                tbl1apk    route  \
0          G4  1.0000     81.43  202131013514112ABEPIE  ABE-PIE   
1          UA  0.1193    154.11  202131013515304ABETPA  ABE-TPA   
2          WN  0.9968    184.44  202131014011259ABQDAL  ABQ-DAL   
3          AA  0.9774    183.09  202131014011298ABQDFW  ABQ-DFW   
4          AA  0.3939    165.77  202131014014107ABQPHX  ABQ-PHX   

                                Geocoded_City1_fixed  \
0  Allentown/Bethlehem/Easton, PA\n(40.602753, -7...   
1  Allentown/Bethlehem/Easton, PA\n(40.602753, -7...   
2          Albuquerque, NM\n(35.084248, -106.649241)   
3          Albuquerque, NM\n(35.084248, -106.649241)   
4          Albuquerque, NM\n(35.084248, -106.649241)   

                                Geocoded_City2_fixed  
0  Tampa, FL (Metropolitan Area)\n(37.8606, -78.8...  
1  Tampa, FL (Metropolitan Area)\n(37.8606, -78.8...  
2      Dallas/Fort Worth, TX\n(40.11086, -77.035636)  
3      Dallas/Fort Worth, TX\n(40.11086, -77.035636)  
4               Phoenix, AZ\n(30.406931, -87.217578)  

[5 rows x 23 columns]

In [22]:
#remove flights with missing values (334)
df.dropna(subset=['Geocoded_City1_fixed'], inplace=True)

In [23]:
#regex to extract coordinates only - standardise
df['coordinates_city1'] = df['Geocoded_City1_fixed'].str.extract(r'(\(\d+.\d+\,\ \-?\d+.\d+\))')
df['coordinates_city2'] = df['Geocoded_City2_fixed'].str.extract(r'(\(\d+.\d+\,\ \-?\d+.\d+\))')

In [24]:
df = df.drop('Geocoded_City1_fixed', axis=1)
df = df.drop('Geocoded_City2_fixed', axis=1)

In [25]:
df.head()

Year  quarter  citymarketid_1  citymarketid_2  \
0  2021        3           30135           33195   
1  2021        3           30135           33195   
2  2021        3           30140           30194   
3  2021        3           30140           30194   
4  2021        3           30140           30466   

                            city1                          city2  airportid_1  \
0  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)        10135   
1  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)        10135   
2                 Albuquerque, NM          Dallas/Fort Worth, TX        10140   
3                 Albuquerque, NM          Dallas/Fort Worth, TX        10140   
4                 Albuquerque, NM                    Phoenix, AZ        10140   

   airportid_2 airport_1 airport_2  ...  carrier_lg  large_ms  fare_lg  \
0        14112       ABE       PIE  ...          G4    1.0000    81.43   
1        15304       ABE       TPA  ...          DL    0.4659   219.98   
2        11259       ABQ       DAL  ...          WN    0.9968   184.44   
3        11298       ABQ       DFW  ...          AA    0.9774   183.09   
4        14107       ABQ       PHX  ...          WN    0.6061   184.49   

  carrier_low   lf_ms  fare_low                tbl1apk    route  \
0          G4  1.0000     81.43  202131013514112ABEPIE  ABE-PIE   
1          UA  0.1193    154.11  202131013515304ABETPA  ABE-TPA   
2          WN  0.9968    184.44  202131014011259ABQDAL  ABQ-DAL   
3          AA  0.9774    183.09  202131014011298ABQDFW  ABQ-DFW   
4          AA  0.3939    165.77  202131014014107ABQPHX  ABQ-PHX   

          coordinates_city1        coordinates_city2  
0   (40.602753, -75.469759)    (37.8606, -78.804199)  
1   (40.602753, -75.469759)    (37.8606, -78.804199)  
2  (35.084248, -106.649241)   (40.11086, -77.035636)  
3  (35.084248, -106.649241)   (40.11086, -77.035636)  
4  (35.084248, -106.649241)  (30.406931, -87.217578)  

[5 rows x 23 columns]

In [26]:
df['airportid-route'] = df[['airportid_1', 'airportid_2']].apply(lambda x: str(x['airportid_1']) + str(x['airportid_2']), axis=1)

In [27]:
# Fix distance column
def replace_with_mode(group):
    mode_distance = group['nsmiles'].mode()[0]
    group['nsmiles'] = mode_distance
    return group

# Apply the function to each group
df = df.groupby(['route']).apply(replace_with_mode)
df = df.reset_index(drop=True)

C:\Users\FTW\AppData\Local\Temp\ipykernel_9940\2248570204.py:8: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df = df.groupby(['route']).apply(replace_with_mode)


In [28]:
# Group by 'route' and aggregate 'nsmiles' to find unique values
grouped = df.groupby('airportid-route')['nsmiles'].unique()

# Find routes with more than one unique mileage
inconsistent_routes = grouped[grouped.apply(len) > 1]

# Count the number of inconsistent routes and rows
num_inconsistent_routes = len(inconsistent_routes)
num_inconsistent_rows = df[df['airportid-route'].isin(inconsistent_routes.index)].shape[0]

print(f"Number of routes with inconsistent mileage: {num_inconsistent_routes}")
print(f"Number of rows with inconsistent mileage: {num_inconsistent_rows}")

Number of routes with inconsistent mileage: 0
Number of rows with inconsistent mileage: 0


In [29]:
df.head()

Year  quarter  citymarketid_1  citymarketid_2  \
0  1994        1           30135           30977   
1  1993        2           30135           30977   
2  1993        3           30135           30977   
3  1993        1           30135           30977   
4  1993        4           30135           30977   

                            city1        city2  airportid_1  airportid_2  \
0  Allentown/Bethlehem/Easton, PA  Chicago, IL        10135        10986   
1  Allentown/Bethlehem/Easton, PA  Chicago, IL        10135        10986   
2  Allentown/Bethlehem/Easton, PA  Chicago, IL        10135        10986   
3  Allentown/Bethlehem/Easton, PA  Chicago, IL        10135        10986   
4  Allentown/Bethlehem/Easton, PA  Chicago, IL        10135        10986   

  airport_1 airport_2  ...  large_ms  fare_lg  carrier_low lf_ms  fare_low  \
0       ABE       CHI  ...      0.80   285.85           CO  0.20    274.45   
1       ABE       CHI  ...      0.84   251.36           US  0.84    251.36   
2       ABE       CHI  ...      0.66   324.72           US  0.66    324.72   
3       ABE       CHI  ...      0.58   246.40           CO  0.25    117.60   
4       ABE       CHI  ...      1.00   343.72           US  1.00    343.72   

                 tbl1apk    route        coordinates_city1  \
0  199411013510986ABECHI  ABE-CHI  (40.602753, -75.469759)   
1  199321013510986ABECHI  ABE-CHI  (40.602753, -75.469759)   
2  199331013510986ABECHI  ABE-CHI  (40.602753, -75.469759)   
3  199311013510986ABECHI  ABE-CHI  (40.602753, -75.469759)   
4  199341013510986ABECHI  ABE-CHI  (40.602753, -75.469759)   

         coordinates_city2 airportid-route  
0  (41.775002, -87.696388)      1013510986  
1  (41.775002, -87.696388)      1013510986  
2  (41.775002, -87.696388)      1013510986  
3  (41.775002, -87.696388)      1013510986  
4  (41.775002, -87.696388)      1013510986  

[5 rows x 24 columns]

In [30]:
# Check if our example inconsistency is fixed
filtered_df = df[df['route'] == 'ABE-PIE']
filtered_df.head()

Year  quarter  citymarketid_1  citymarketid_2  \
40  2021        3           30135           33195   
41  2022        2           30135           33195   
42  2008        1           30135           33195   
43  2007        1           30135           33195   
44  2007        3           30135           33195   

                             city1                          city2  \
40  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)   
41  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)   
42  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)   
43  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)   
44  Allentown/Bethlehem/Easton, PA  Tampa, FL (Metropolitan Area)   

    airportid_1  airportid_2 airport_1 airport_2  ...  large_ms  fare_lg  \
40        10135        14112       ABE       PIE  ...       1.0    81.43   
41        10135        14112       ABE       PIE  ...       1.0    98.33   
42        10135        14112       ABE       PIE  ...       1.0   108.71   
43        10135        14112       ABE       PIE  ...       1.0   104.83   
44        10135        14112       ABE       PIE  ...       1.0    91.23   

    carrier_low lf_ms  fare_low                tbl1apk    route  \
40           G4   1.0     81.43  202131013514112ABEPIE  ABE-PIE   
41           G4   1.0     98.33  202221013514112ABEPIE  ABE-PIE   
42           G4   1.0    108.71  200811013514112ABEPIE  ABE-PIE   
43           G4   1.0    104.83  200711013514112ABEPIE  ABE-PIE   
44           G4   1.0     91.23  200731013514112ABEPIE  ABE-PIE   

          coordinates_city1      coordinates_city2 airportid-route  
40  (40.602753, -75.469759)  (37.8606, -78.804199)      1013514112  
41  (40.602753, -75.469759)  (37.8606, -78.804199)      1013514112  
42  (40.602753, -75.469759)  (37.8606, -78.804199)      1013514112  
43  (40.602753, -75.469759)  (37.8606, -78.804199)      1013514112  
44  (40.602753, -75.469759)  (37.8606, -78.804199)      1013514112  

[5 rows x 24 columns]

In [31]:
#export cleaned csv
df.to_csv('us_airplane_dataset_cleaned.csv', index=False)